api/utils/sshutils/ppk/ppk.go

```diff
@@ -0,0 +1,223 @@
+	binary.Write(macInput, binary.BigEndian, macEncryptionType)
+
+	// string: the key-comment-string header field.
+	macComment := getRFC4251String([]byte(fileComment))
+	binary.Write(macInput, binary.BigEndian, macComment)
+
+	// base64-decode the Teleport public key, as we need its binary representation to generate the MAC
+	teleportPublicKeyDecoded := make([]byte, base64.StdEncoding.EncodedLen(len(pub)))
+	teleportPublicKeyByteCount, err := base64.StdEncoding.Decode(teleportPublicKeyDecoded, pub)
+	if err != nil {
+		return nil, trace.Errorf("could not base64-decode public key: %v, got %v bytes successfully", err, teleportPublicKeyByteCount)
+	}
+	publicKeyData := make([]byte, teleportPublicKeyByteCount)
+	copy(publicKeyData, teleportPublicKeyDecoded)
+	// append the decoded public key bytes to the MAC buffer
+	macPublicKeyData := getRFC4251String(publicKeyData)
+	binary.Write(macInput, binary.BigEndian, macPublicKeyData)
+
+	// append our PPK-formatted private key bytes to the MAC buffer
+	macPrivateKeyData := getRFC4251String(ppkPrivateKey.Bytes())
+	binary.Write(macInput, binary.BigEndian, macPrivateKeyData)
+
+	// as per the PPK spec, the key for the MAC is blank when the PPK file is unencrypted.
+	// therefore, the key is a zero-length byte slice.
+	hmacHash := hmac.New(sha256.New, []byte{})
+	// generate the MAC using HMAC-SHA-256
+	hmacHash.Write(macInput.Bytes())
+	macString := hex.EncodeToString(hmacHash.Sum(nil))
+
+	// build the string-formatted output PPK file
+	ppk := new(bytes.Buffer)
+	fmt.Fprintf(ppk, "PuTTY-User-Key-File-3: %v\n", keyType)
+	fmt.Fprintf(ppk, "Encryption: %v\n", encryptionType)
+	fmt.Fprintf(ppk, "Comment: %v\n", fileComment)
+	// chunk the Teleport-formatted public key into 64-character length lines
+	chunkedPublicKey := chunkBase64String(string(pub), 64)
+	fmt.Fprintf(ppk, "Public-Lines: %v\n", len(chunkedPublicKey))
+	for _, r := range chunkedPublicKey {
+		fmt.Fprintf(ppk, "%s\n", r)
+	}
+	// chunk the PPK-formatted private key into 64-character length lines
+	chunkedPrivateKey := chunkBase64String(string(ppkPrivateKeyBase64), 64)
+	fmt.Fprintf(ppk, "Private-Lines: %v\n", len(chunkedPrivateKey))
+	for _, r := range chunkedPrivateKey {
+		fmt.Fprintf(ppk, "%s\n", r)
+	}
+	fmt.Fprintf(ppk, "Private-MAC: %v\n", macString)
+
+	return ppk.Bytes(), nil
+}
+
+// chunkBase64String converts a string into a []string with chunks of size chunkSize;
+// used to split base64-encoded strings across multiple lines with an even width.
+// note: this function operates on Unicode code points rather than bytes, therefore
+// using it with multi-byte characters will result in unevenly chunked strings.
+// it's intended usage is only for chunking base64-encoded strings.
+func chunkBase64String(s string, chunkSize int) []string {
+	if len(s) == 0 {
+		return nil
+	}
+	if chunkSize >= len(s) {
+		return []string{s}
+	}
+	chunks := make([]string, 0, (len(s)-1)/chunkSize+1)
+	currentLen := 0
+	currentStart := 0
+	for i := range s {
```

Comment:
This loop seems a bit awkward, I don't think you need to iterate over each byte and increment a var each time. You basically just want to take the min of chunkSize or whatever is left. How about something like:

```
func chunk(s string, size int) []string {
	var chunks []string
	for b := []byte(s); len(b) > 0; {
		n := size
		if n > len(b) {
			n = len(b)
		}
		chunks = append(chunks, string(b[:n]))
		b = b[n:]
	}
	return chunks
}
```